src/Middleware/HttpLogging/src/FileLoggerProcessor.cs

```diff
@@ -0,0 +1,262 @@
+                    {
+                        _fileNumber++;
+                        fullName = GetFullName(today);
+                        fileInfo = new FileInfo(fullName);
+                        streamWriter.Dispose();
+                        if (!TryCreateDirectory())
+                        {
+                            // return early if we fail to create the directory
+                            return;
+                        }
+                        streamWriter = GetStreamWriter(fullName);
+                    }
+                    if (!fileInfo.Exists || fileInfo.Length == 0)
+                    {
+                        await OnFirstWrite(streamWriter, _cancellationTokenSource);
+                    }
+
+                    await WriteMessageAsync(message, streamWriter, _cancellationTokenSource);
+                }
+            }
+            finally
+            {
+                RollFiles();
+                streamWriter.Dispose();
+            }
+
+        }
+
+        internal bool TryCreateDirectory()
+        {
+            if (!Directory.Exists(_path))
+            {
+                try
+                {
+                    Directory.CreateDirectory(_path);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    Log.CreateDirectoryFailed(_logger, _path, ex);
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        // Virtual for testing
+        internal virtual async Task WriteMessageAsync(string message, StreamWriter streamWriter, CancellationTokenSource cancellationTokenSource)
+        {
+            if (cancellationTokenSource.IsCancellationRequested)
+            {
+                return;
+            }
+            await streamWriter.WriteLineAsync(message);
+            await streamWriter.FlushAsync();
+        }
+
+        // Virtual for testing
+        internal virtual StreamWriter GetStreamWriter(string fileName)
+        {
+            return File.AppendText(fileName);
+        }
+
+        private void RollFiles()
+        {
+            if (_maxRetainedFiles > 0)
+            {
+                var files = new DirectoryInfo(_path)
+                    .GetFiles(_fileName + "*")
+                    .OrderByDescending(f => f.Name)
+                    .Skip(_maxRetainedFiles.Value);
+
+                foreach (var item in files)
+                {
+                    item.Delete();
+                }
+            }
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            _cancellationTokenSource.Cancel();
+            _messageQueue.CompleteAdding();
+            await _outputTask;
+        }
+
+        private string GetFullName(DateTime date)
+        {
+            return Path.Combine(_path, FormattableString.Invariant($"{_fileName}{date.Year:0000}{date.Month:00}{date.Day:00}.{_fileNumber:0000}.txt"));
```

Comment:
@Tratcher I went back to a 4-digit number for fileNumber - with an arbitrary length fileNumber, sorting files by name would think "file.1" was around as old as "file.1000", so file rolling wasn't actually deleting the oldest files.

Reply:
This and the `OnChange` callback should both be locked so `_path` and `_fileName` never go out of sync.